M2/Macaulay2/packages/Style.m2

```diff
@@ -3,16 +3,184 @@ newPackage( "Style",
      AuxiliaryFiles => true,
      Headline => "style sheets and images for the documentation",
      Keywords => {"Documentation"},
-     Version => "1.0"
+     Version => "1.1"
      )
 
+---------------------------------------------
+-- grammar generation code adapted from    --
+-- M2/Macaulay2/editors/make-M2-symbols.m2 --
+---------------------------------------------
+
+export {"generateGrammar"}
+
+importFrom(Core, "sortBy")
+
+-- TODO: Move these two elsewhere:
+Function and Function := (f, g) -> s -> f s and g s
+Function or  Function := (f, g) -> s -> f s or  g s
+
+is := X -> (name, symb) -> instance(value symb, X)
+
+isAlpha        := s -> match("^[[:alpha:]]+$", s)
+isAlphaNumeric := s -> match("^[[:alnum:]]+$", s)
+
+-- Things we want to highlight:
+isType     := is Type
+isKeyword  := is Keyword
+isFunction := is Function
+isConst    := (name, symb) -> (isAlphaNumeric name
+    and not (isFunction or isType or isKeyword) (name, symb)
+    and (symb === symbol null or value symb =!= null))
+
+okay := method()
+okay(String, Keyword) :=
+okay(String, Symbol)  := (name, pkg) -> length name > 1 and isAlphaNumeric name
+
+cachedSymbols = new CacheTable
+STRINGS = format "///\\\\(/?/?[^/]\\\\|\\\\(//\\\\)*////[^/]\\\\)*\\\\(//\\\\)*///"
+
+-- This banner is added to the top of generated grammars
+banner := "Auto-generated for Macaulay2-@M2VERSION@. Do not modify this file manually."
+
+generateSymbols = () -> (
+    ----------------------------------------------------------------------------
+    -- Get a list of all symbols visible just after loading preloaded packages
+    allPkgNames := separate(" ", version#"packages") | {"Core"};
+    loadedPkgNames := join(Core#"preloaded packages",
+	{"Core", "Text", "Parsing", "SimpleDoc"});
+    symbols := unique (sortBy ascii @@ first) join(
+	apply(allPkgNames, pkgname -> (pkgname, symbol Core)),
+	flatten apply(loadedPkgNames, pkgname -> (
+		pkg := needsPackage pkgname;
+		select(pairs pkg.Dictionary, okay))));
+
+    if length symbols < 1500 then error "expected more entries for M2-symbols";
+
+    -- Check for invalid symbols
+    bad := select(symbols, (name, symb) -> not okay(name, symb));
+    if #bad > 0 then error(
+	"encountered symbol(s) that are not alphanumeric or have less than 2 characters:",
+	concatenate apply(bad, (name, symb) -> {
+		"\n\t",
+		-* TODO: symbolLocation name, ": here is the first use of ", *-
+		toString name}));
```

Comment:
I recently learned something super cool about how `error` prints symbols which I didn't know when I wrote this:
```m2
i7 : error splice("encountered symbol(s) that are not alphanumeric or have less than 2 characters:",
     toSequence between_" " {symbol Module, symbol print})
stdio:13:5:(3): error: encountered symbol(s) that are not alphanumeric or have less than 2 characters:'Module' 'print'
../linuxbrew/.linuxbrew/share/Macaulay2/Core/expressions.m2:1234:42-1234:48: here is the first use of 'Module'
../linuxbrew/.linuxbrew/share/Macaulay2/Core/expressions.m2:1158:0-1158:5: here is the first use of 'print'
```

Reply:
That's neat!

I'm a little confused at this whole error message, though.  The `symbols` list appears to be made up of two things: the list of all packages and everything from the dictionaries in the pre-loaded packages.  We already call `select` with `okay` on the latter, so we know that none of those will end up in `bad`.  Why not just do the same thing on the former?

And also, these packages are always paired up with the `Core` symbol (I guess so that when we call `is` on them, we know that they're packages), so even if one of them ends up in `bad` and we modify the error message to print the symbol location, it will just be the location of `Core`.

Reply:
I'm not sure. Maybe experiment with not selecting on line 51 and see if anything breaks?

Reply:
There are *tons* of non-alphanumeric symbols (mathematical operators, all the `Package$foo` ones) that were getting filtered out by `select`.

The only elements of this list we weren't selecting out with `okay` are the package names, and we know that they'll be alphanumeric since `newPackage` raises an error when they're not.  I just pushed a commit removing this check.